akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -97,12 +97,8 @@ private[stream] class ConnectionSourceStage(val tcpManager: ActorRef,
         connectionFlowsAwaitingInitialization.incrementAndGet()
 
         val tcpFlow =
-          Flow.fromGraph(new IncomingConnectionStage(connection, connected.remoteAddress, halfClose))
+          Flow.fromGraph(new IncomingConnectionStage(connection, connected.remoteAddress, halfClose, connectionFlowsAwaitingInitialization))
             .via(detacher[ByteString]) // must read ahead for proper completions
```

Comment:
the detacher? Keeping it as an extra stage makes the TCP stages *much* simpler and easier to understand.